cookbooks/creating_a_behat_extension.rst

```diff
@@ -0,0 +1,285 @@
+          $this->filepath = $filepath;
+          $this->enable = $enable;
+      }
+  
+      public function initializeContext(Context $context)
+      {
+          /* 
+           * At the start of every scenario, behat will create a new instance of every `Context`
+           * registered in your project. It will then call this method with each new `Context` in
+           * turn. If you want to initialise multiple contexts, you can of course give them an 
+           * interface and check for that here.
+           */
+          if (!$context instanceof LogsContext) {
+              return;
+          }
+
+          $context->initializeConfig($this->enable, $this->filepath);
+      }
+  }
+
+We need to register the initializer definition within the Behat container through the ``LogsExtension``, ensuring it gets loaded:
+
+.. code-block:: php
+
+  <?php
+
+  // ...
+
+  use Symfony\Component\DependencyInjection\Definition;
+  use Behat\Behat\Context\ServiceContainer\ContextExtension;
+
+  class LogsExtension implements Extension
+  {
+      // ...
+  
+      public function load(ContainerBuilder $container, array $config)
+      {
+          $definition = new Definition(LogsInitializer::class, [
+              $config['filepath'],
+              $config['enable'],
+          ]);
+          $definition->addTag(ContextExtension::INITIALIZER_TAG);
+          $container->setDefinition('logs_extension.context_initializer', $definition);
+      }
+
+      // ...
+  }
+
+To complete the extension, we must add methods to ``LogsContext`` to receive the configuration values and use those in the hooks:
+
+.. code-block:: php
+
+  // ...
+  
+  class LogsContext implements Context
+  {
+      private bool $enable = false;
+      private string $filepath;
+  
+      public function initializeConfig(bool $enable, string $filepath)
+      {
+          $this->enable = $enable;
+          $this->filepath = $filepath;
+      }
+  
+      /** @BeforeScenario */
+      public function before(BeforeScenarioScope $scope)
+      {
+          if ($this->enable === false) {
+              return;
+          }
+
+          file_put_contents(
+              $this->filepath,
+              'START: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
+              FILE_APPEND
+          );
+      }
+  
+      /** @AfterScenario */
+      public function after(AfterScenarioScope $scope)
+      {
+          if ($this->enable === false) {
+              return;
+          }
+  
+          file_put_contents(
+              $this->filepath,
+              'END: ' . $scope->getScenario()->getTitle() . ' - ' . time() . PHP_EOL,
+              FILE_APPEND
+          );
+      }
+  }
+
+Conclusion
+----------
+
+Congratulations! You have just created a simple Behat extension that logs scenario durations. This extension demonstrates the three essential steps to building a Behat extension: defining an extension, creating an initializer, and configuring contexts.
```

Comment:
Using a context to log things is probably a bad example. It won't work out of the box, as your context needs to be registered in the config of the project to be used (which is not even described in your cookbook).
The extension should use events instead of hooks instead.

Reply:
> The extension should use events instead of hooks instead.
There's another way to access events apart hooks?
Have you got an example or a link to share?

Btw, I just push a new commit with an example of a `behat.yaml` configuration: f91f7c9

Reply:
@monitaurus yes - you can instead [register an event subscriber](https://github.com/ingenerator/behat-support/blob/66028a7f3ed4195a3c042bc8e49966a197753144/src/Extension/SaveFailingPagesExtension.php#L49-L58) and [implement the EventSubscriberInterface](https://github.com/ingenerator/behat-support/blob/1.x/src/Extension/SaveFailingPagesExtension/SaveFailingPagesListener.php) - I'm not saying those are good examples overall BTW, I just knew where to find them quickly :laughing: 

There are a few advantages, partly your behat.yml doesn't have to register a context as well as an extension, partly that way you can set the priority of your subscriber so you can control when it runs relative to hooks and other listeners.

**HOWEVER** since this is a cookbook I would also be OK with keeping your example how it is, now that it mentions that there are other ways to do this and has a behat.yml showing how it works.

Your example has the advantage that:

* It also shows how an extension can initialise a context.
* It minimises the number of "new things" people need to consider. They have probably already written hooks in their own contexts and understand how they work. For example, the event listeners get slightly different payloads, and with the power to set priority also comes the ability to cause problems that don't exist with hooks.

And it works, so long as you put the right config in behat.yml.

So my temptation is to say as long as it's clear we're saying this is a simplified example, and not the only / best way (which I think it does) then this is useful advice for people.

Reply:
A cookbook showing how an extension can initialize a context is indeed worth it. But then I would suggest 2 changes:

- not naming it `Creating a Behat extension`, as this cookbook is about initializing a context from an extension, which is only a specific thing an extension can do
- maybe find a better example for what the context is doing (it should probably be a context providing some step definition, to be a case where this is indeed the right tool to solve the need)

Reply:
_Oh my 😱 I didn't knew that.
At work I made a OpenAPI extension that I need to load the context every time.
I'll change it to use subscriber._

I agree @stof, for a log extension, a subscriber is far better.
But as @acoulton say, as it's a cookbook, I think we can leave it this way.
I also fear it may overload the text with another notion.

I suggest a direct follow up to this cookbook, a _"Subscribing to events"_.
It could take the same boilerplate as this cookbook, and make the step further to change hooks to subscribers.

Reply:
Ok, @stof @acoulton, I change the extension example to a simple Hello World.
The extension give a Context which will throw a PendingException which message is configurable in parameters.

The extension is clearly a nonsense, but allow to simplify a lot the whole cookbook by removing the hook.

What do you think?